Add PrimitiveSerializer and StructuredSerializer interfaces.

davidmorgan opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/built_json.dart/20)

<!-- Reviewable:end -->

davidmorgan committed:
Add PrimitiveSerializer and StructuredSerializer interfaces.

BlackHC commented:
Review status: 0 of 20 files reviewed at latest revision, 4 unresolved discussions.

---

_[built_json/lib/built_json.dart, line 155 [r1]](https://reviewable.io:443/reviews/google/built_json.dart/20#-KD3U2rhPY_Q4lFwmIF9:-KD3U2rifLsYWH9xwNgE:548993694) ([raw file](https://github.com/google/built_json.dart/blob/551ca0f05f5bfe945dbac1a6f1080d921fe9b807/built_json/lib/built_json.dart#L155)):_
Is specifiedType needed for primitive types? 

The doc comments need to be updated for both classes btw.  This one is for primitive types (as defined in Dart) and the one below for composite types or whatever the right term is ^^

---

_[built_json/lib/src/built_json_serializers.dart, line 128 [r1]](https://reviewable.io:443/reviews/google/built_json.dart/20#-KD3UkVBS-8mKrKtq1Cl:-KD3UkVBS-8mKrKtq1Cm:-2135743432) ([raw file](https://github.com/google/built_json.dart/blob/551ca0f05f5bfe945dbac1a6f1080d921fe9b807/built_json/lib/src/built_json_serializers.dart#L128)):_
Is this necessary? Sorry for nitpicking. (Could we also just use marker interfaces?)

---

_[built_json/lib/src/built_map_serializer.dart, line 60 [r1]](https://reviewable.io:443/reviews/google/built_json.dart/20#-KD3Vs0ouWjUZRjJyvv2:-KD3Vs0ouWjUZRjJyvv3:-1549607517) ([raw file](https://github.com/google/built_json.dart/blob/551ca0f05f5bfe945dbac1a6f1080d921fe9b807/built_json/lib/src/built_map_serializer.dart#L60)):_
Bit magic? Why not % 2?

---

_[built_json/lib/src/built_set_serializer.dart, line 40 [r1]](https://reviewable.io:443/reviews/google/built_json.dart/20#-KD3W0YKiztQ0u16MgCs:-KD3W0YKiztQ0u16MgCt:-943958279) ([raw file](https://github.com/google/built_json.dart/blob/551ca0f05f5bfe945dbac1a6f1080d921fe9b807/built_json/lib/src/built_set_serializer.dart#L40)):_
Why? (readability)

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/google/built_json.dart/20)_

<!-- Sent from Reviewable.io -->

davidmorgan commented:
Review status: 0 of 20 files reviewed at latest revision, 4 unresolved discussions.

---

_[built_json/lib/built_json.dart, line 155 [r1]](https://reviewable.io:443/reviews/google/built_json.dart/20#-KD3U2rhPY_Q4lFwmIF9:-KD4A3DisEUTEathYX1w:-2099197343) ([raw file](https://github.com/google/built_json.dart/blob/551ca0f05f5bfe945dbac1a6f1080d921fe9b807/built_json/lib/built_json.dart#L155)):_
It's not used right now, but it's conceivable, particularly if people want to plug in other serialization systems. e.g. if you wanted something that serializes everything to a single String, it would be a PrimitiveSerializer but would care about generics.

Re: the doc comments, yeah, that was a big ugly. There is really no inheritance for serialize/deserialize, they're disjoint. So I've moved them onto the subclasses and added separate docs. The downside is it means I have to add another branch to the "if" clause in built_json_serializers: if (is StructuredSerializer) else if (is PrimitiveSerializer) else {}. This is clearer though. Done.

---

_[built_json/lib/src/built_json_serializers.dart, line 128 [r1]](https://reviewable.io:443/reviews/google/built_json.dart/20#-KD3UkVBS-8mKrKtq1Cl:-KD4HVmzvA1Unz50VzgS:120867991) ([raw file](https://github.com/google/built_json.dart/blob/551ca0f05f5bfe945dbac1a6f1080d921fe9b807/built_json/lib/src/built_json_serializers.dart#L128)):_
Yeah, I think we need both types, now that I moved the methods to where they should really be the top level "serializer" class can't serialize or deserialize anything. But, we still need that interface so we can find the right serializer. PTAL.

---

_[built_json/lib/src/built_map_serializer.dart, line 60 [r1]](https://reviewable.io:443/reviews/google/built_json.dart/20#-KD3Vs0ouWjUZRjJyvv2:-KD4HpHJVjSwwN_F7qsc:-20274394) ([raw file](https://github.com/google/built_json.dart/blob/551ca0f05f5bfe945dbac1a6f1080d921fe9b807/built_json/lib/src/built_map_serializer.dart#L60)):_
Dunno, for me it reads the same :) ... changed.

---

_[built_json/lib/src/built_set_serializer.dart, line 40 [r1]](https://reviewable.io:443/reviews/google/built_json.dart/20#-KD3W0YKiztQ0u16MgCs:-KD4IK7GHmuwGxaap_Nx:-77884384) ([raw file](https://github.com/google/built_json.dart/blob/551ca0f05f5bfe945dbac1a6f1080d921fe9b807/built_json/lib/src/built_set_serializer.dart#L40)):_
Underspecified means there's no information about the types it contains, so we have to create a Set<Object>.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/google/built_json.dart/20)_

<!-- Sent from Reviewable.io -->

BlackHC commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 20 files reviewed at latest revision, all discussions resolved.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/google/built_json.dart/20#-:-KD4LrBTkltHjms4-Nvn:1417088721)_

<!-- Sent from Reviewable.io -->
